### ISSUE TITLE
add support for /api/node/{type} and/or /api/v2/node/{type}

### ISSUE THREAD

**Issue by ps23Rick** [backdrop-contrib/headless#41]
I know that other JSON implementations (e.g. the Drupal one I used recently) allows querying all nodes of a specified type without the need to specify any {id} values.  This will return, in these implementations, all the resulting node-ids of the specified type.  This is currently missing and I'm willing to add the support if it's of interest. Also, if there's an alternative way to achieve this through the existing support, I'm open to using that.

For instance, my code that currently works with Drupal can query all nodes of custom type {navpage} using the Drupal API : `http://localhost:8888/drupal/jsonapi/node/navpage` and then it can weed through the results.  I should note that in this case the result set is always very small and no pagination is required. 

I tried defining a backdrop view for this type {navpage}, and told it to display everything, no pagination, no menu, unformatted, unsorted.  I can pull that view up just fine and see the content using only the backdrop URL, but when I enable the respective view within the headless module, it just returns this terse information as shown below :

`{"results":[{"nid":"3"}],"count":1}`

I was hoping it would return the entire content block for that particular navpage item..I looked at the code for the views module and it apparently only returns those items.  Hmm...

----

PR: https://github.com/backdrop-contrib/headless/pull/42

**Comment by serundeputy** [comment]
Hello @ps23Rick !

If the view had fields on it it would return the node fields.  That is how i always use it in conjuction w/ headless module.

This page: https://api.serundeputy.io/api/v2/views/archives_api/page
provides the data for this page: https://serundeputy.io/archives

for example.

That said I'm not opposed to adding collection support for the types the user opt into the API.

**Comment by ps23Rick** [comment]
Hmm.. I went back and reviewed my "view" settings and you were right -- I had chosen unformatted output and that disabled the fields setting.  I forced it to be turned on but still didn't get anything meaningful for some reason.  With that said, assuming that you're OK with me doing so, wouldn't mind implementing either one of the following APIs as an extension of the existing node handling:

`api/node/{type}`

OR 

`api/nodes/{type}`

I believe I've seen both implemented as part of different JSON implementations for different CMS's.  Just let me know what you want to do..  If you're OK with moving forward, I'll try my best to tackle it today.

**Comment by serundeputy** [comment]
I'm fine w/ you moving forward on this. I think i'd use `api/node/{type}`.

**Comment by ps23Rick** [comment]
ok.. I'll proceed then using api/node/{type}.

**Comment by ps23Rick** [comment]
ok.. stupid question to ponder.  

In the cases where a bunch of nodes can be returned, do you want to do pages/ranges of the resulting data?  This is not really an issue for me as my datasets will be very small -- like probably under 10.  However, I realize that others may have much larger data sets.  Currently my initial code will gather up to 10 items from the database before returning.  I've not bothered looking through the other parts of the headless code to see if it handles pagination or just grabs whatever is there.  Maybe the existing code never had to deal with this scenario.  Thoughts?

**Comment by serundeputy** [comment]
The views endpoints handle pagination, but that might be from the view query and not the api query.

**Comment by serundeputy** [comment]
yeah; looks like the pagination is handled on the views end for `headless_views_v2`.

So you'll need to include pagination in your sql query.

**Comment by ps23Rick** [comment]
ok.. Initially I'll keep things simple just to ensure the underlying queries are working and being provided back as JSON and will implement the pagination after the initial stuff is working as expected.  Agreed. I just looked at what you indicated and see the pagination code (`_views_get_view_result`) referenced

**Comment by serundeputy** [comment]
I think something like this will work

```php
$sql = db_select('node', 'n')->extend('PagerDefault');
$nodes = $sql
  ->fields('n', array('nid'))
  ->limit(4)
  ->execute();

foreach ($nodes as $node) {
  print_r($node);
}

```

**Comment by ps23Rick** [comment]
Right now I'm using the following code to pull up the list of nodes of {type} and it appears to be working quite well -- I found it as part of a search.. 

```
  // load the requested node and check to see if types match or not
  $query = new EntityFieldQuery();
  $query->entityCondition('entity_type', 'node')
        ->entityCondition('bundle', $type)
        ->propertyCondition('status', 1) // get only published items
      // It returns the first 10 nodes.
      // Use range(10, 10) to get the next 10 nodes.
        ->range(0, 10);  
  $result = $query->execute();
  if (isset($result['node'])) {
    $nodes = entity_load('node', array_keys($result['node']));
  }

```

Is one better than the other?  I've got no idea frankly...

**Comment by serundeputy** [comment]
You can also use `devel_generate` if you want to generate 100s or 1000s of nodes for testing.

```bash
drush dl devel
```

```bash
drush en devel_generate
```

```bash
drush help genc
```

**Comment by serundeputy** [comment]
`EntityFieldQuery` is fine.

**Comment by ps23Rick** [comment]
thanks.. I'll see how far I can get -- the pagination complicates things a little.  it's not likely to be done today though.. So much for my running streak!  lol

**Comment by serundeputy** [comment]
for pagination we can use url params like `/api/v2/node/page?page=2` etc default to `page=1`;
https://api.backdropcms.org/api/backdrop/core%21includes%21common.inc/function/backdrop_get_query_array/1

**Comment by ps23Rick** [comment]
ok.. Stupid question.. I'm back working on this update again and I'm trying to implement the v2 version of the code.  I added my two new table entries in the headless_menu() function :

```
  // The following is for the node type where an id is not specified - to return all of the specified
  $items['api/node/%'] = array(
    'page callback' => 'headless_types',
    'access callback' => TRUE,
    'page arguments' => array(2),
  );
  // The following is for the node type where an id is not specified - to return all of the specified
  $items['api/v2/node/%'] = array(
    'page callback' => 'headless_types_v2',
    'access callback' => TRUE,
    'page arguments' => array(2),
  );
```

The first one works fine -- the code is invoked, but the 2nd one (the v2) comes up with a page not found when I try to load "/api/v2/node/{type}".  What am I missing in the plumbing to get things so that my new function can get called? I'm sure I'm missing something.  I checked the boxes in the admin page for the headless module to enable the v2 functionality -- but not sure it's quite right.

ok.. I tried this URL in Insomnia: -- "navpage" is the type I'm using for my testing.

`http://localhost/api/v2/node/navpage/page?page=1`

with that specified, I get a 404. -- I'm sure my code is not being invoked.

**Comment by indigoxela** [comment]
> 'page arguments' => array(2),

The second item needs `array(3)` as the placeholder is the 4th path part (arguments start with number 0).

**Comment by ps23Rick** [comment]
aaah.. Thanks.. I hadn't thought of that. Thanks so much!

**Comment by ps23Rick** [comment]
Ok.. I must be missing something.. Its still not working.  Below are all 4 of the table entries for the JSON nodes :

```
  // Return JSON for individual nodes.
  $items['api/node/%/%'] = array(
    'page callback' => 'headless_type',
    'access callback' => TRUE,
    'page arguments' => array(2, 3),
  );
  $items['api/v2/node/%/%'] = array(
    'page callback' => 'headless_type_v2',
    'access callback' => TRUE,
    'page arguments' => array(3, 4),
  );
  // The following is for the node type where an id is not specified - to return all of the specified
  $items['api/node/%'] = array(
    'page callback' => 'headless_types',
    'access callback' => TRUE,
    'page arguments' => array(2),
  );
  // The following is for the node type where an id is not specified - to return all of the specified
  $items['api/v2/node/%'] = array(
    'page callback' => 'headless_types_v2',
    'access callback' => TRUE,
    'page arguments' => array(3),
  );

```

Is it possible that because I've got 4 entries for nodes now instead of 2 that the "routing" (for lack of a better term) is being confused?  I can get the first 3 to work but the last one refuses and comes up with the following.

<img width="495" alt="Screen Shot 2020-04-23 at 11 19 08 AM" src="https://user-images.githubusercontent.com/38017680/80134968-6535e780-8554-11ea-9996-4bd7b65b23ed.png">

Below is part of the page configuration for headless.. I don't know how to get it to show that there are now 4 possible options for pulling up nodes instead of 2.. 

<img width="1108" alt="Screen Shot 2020-04-23 at 11 23 55 AM" src="https://user-images.githubusercontent.com/38017680/80135422-0fae0a80-8555-11ea-8dfa-b4527c4ab5d6.png">

If you've got any thoughts.. ??

**Comment by serundeputy** [comment]
@ps23Rick not sure about the routing part at a glance things seem in order

for the settings page table; looks like it is being built here: https://github.com/backdrop-contrib/headless/blob/1.x-1.x/includes/headless.admin.inc#L42

If you want to push up a PR w/ what you have I can pull it down and try to investigate the routing issue.

**Comment by ps23Rick** [comment]
ok.. Give me a few minutes to recall the steps to do that.. lol.. thx!!

**Comment by ps23Rick** [comment]
ok.. pr created.

**Comment by serundeputy** [comment]
thanks @ps23Rick I left some comments on the PR.

**Comment by ps23Rick** [comment]
ok.. I'm playing with this code again -- I'm still stuck with the routing issue.  I've put debug statements in the file :

```
function headless_router(string $path) {

      printf("foo12!!!\n");
...

```

Unfortunately I do not see this being handled -- nothing.  Am I missing something?

**Comment by serundeputy** [comment]
sanity check ... did you clear cache after adding the new functions? 

UI: `admin/config/development/performance`
CLI: `drush cc all`

caches need to be cleared for Backdrop to pick up new files, functions, or classes.

**Comment by ps23Rick** [comment]
Doh!  That was EXACTLY my problem.. I cleared the caches and now things are moving again!!!  Thanks!  Now I can move forward on this thing!

**Comment by ps23Rick** [comment]
Stupid question.. In looking at the code for _views_get_view_result(), that method calls a backdrop API method called  views_get_view().  In following a similar pattern for the nodes w/ pagination, I can't seem to find an equivalent nodes function but I think node_get_node_tempstore() might be an equivalent..  Is that about right or should I not even bother to try to keep the pattern similar?  Just thought I'd get your $0.02 worth.

**Comment by serundeputy** [comment]
maybe this is what you are after https://api.backdropcms.org/api/backdrop/core%21modules%21node%21node.module/function/node_load_multiple/1 , but i'm not sure you need that in this case i'd have to look at your code more closely.

**Comment by ps23Rick** [comment]
ok.. I'm back after I spent quite some time today getting my docker image squared away and usable with xDebug and PHPStorm.  

So.. in terms of paginating this data, I see that the existing view code does something along those lines but the pagination appears to be handled outside this codebase if I'm not too far off track.  Unfortunately I will need to do my own thing here as far as I can tell. 

With that said, we know that the JSON standard supports pagination obviously, but my question revolves around the syntax. I know that above you indicated that I should handle the case :

`GET /api/v2/node/{type}?page=2`

whereas the spec suggests :
`GET /api/v2/node/{type}?page[number]=2`

I just want to ensure that I'm implementing what you desire in this area.. So, I'm going to assume the first is what you want.  **Please confirm.**

The next question has to do with pagesize.. Do you want to support page sizes being specified by the user?  eg. the spec claims it should support stuff like the following:

`GET /api/v2/node/{type}?page[number]=2&page[size]=15`

IF you want to support that AND you want the syntax to be similar to the above, I could implement a query similar to the following:

`GET /api/v2/node/{type}?page=2&size=15`

IF you don't want to support page sizing directly we could just hard-code the page size to be 10 or some other constant.  Thoughts??

**Comment by serundeputy** [comment]
the syntax i suggested `GET /api/v2/node/{type}?page=2` is because that is how views does it, but I'd prefer to stick w/ the jsonapi spec as that is widely adopted.

I think it would be better more usefull if the user can specify the page size, but if it is easier to hard code we could just get that in and iterate later to include user specified page size.

thanks for sticking w/ this!

**Comment by ps23Rick** [comment]
Ok.. I’ll follow the jsonapi spec.. I’ll try my best to follow this out and include the page size as well. Thx! Using a full debugger makes life oh so good again!! 



Sent from myMail for iOS


Tuesday, May 19, 2020, 7:43 PM -0700 from Geoff St. Pierre  <notifications@github.com>:

**Comment by ps23Rick** [comment]
ok.. quick question for ya.. 

my router code for mapping the request to my function looks as follows :

```// The following is for the node type where an id is not specified - to return all of the specified
  $items['api/v2/node/%'] = array(
    'page callback' => 'headless_types_v2',
    'access callback' => TRUE,
    'page arguments' => array(3),
  );
```
This works fine except for the extra args that we're discussing (page,size).  Those show up as GET values but not arguments to the function.  Is there some better/best way to add these values into my function in a standardized way.. Obviously I can dig them out of the GET values directly but if there's a better way, I'm all ears. Thanks!

**Comment by serundeputy** [comment]
use `backdrop_get_query_parameters()` https://api.backdropcms.org/api/backdrop/core%21includes%21common.inc/function/backdrop_get_query_parameters/1

**Comment by ps23Rick** [comment]
Thanks for that.  things are progressing.  One thing I wanted to note.. In the case where the query comes in and the page size hasn't been specified but the page number has.. I'm thinking of using a constant in this case to assume a specified page size.  

Or would you like to force that both the page size and page number always be provided as a pair and IF the user specifies one without the other that an error get pitched back..?  Thx!

**Comment by serundeputy** [comment]
> Or would you like to force that both the page size and page number always be provided as a pair and IF the user specifies one without the other that an error get pitched back..? Thx!

I think if they don't specify a page size default to 10 or 20 seems reasonable.

**Comment by ps23Rick** [comment]
Hmm. I was just reading the spec and it seems to imply that they both should be provided together.  I can go either way.. IF you want it to force the client to specify both always, then in the case where only one is specified, I can have the code generate a 422 error or similar and indicate why...

**Comment by serundeputy** [comment]
well the lovely folks behind jsonapi were kind enough to provide us a spec, so let's follow it.

**Comment by ps23Rick** [comment]
ok.. I've got it working but will do additional testing & cleanup.

One thing I wanted to check on.. There are, as with all the headless stuff, two methods -- a plain and a "v2".  The pagination is completely and ONLY in the v2 code.  There is no pagination in the vanilla code and if you pass those new optional args in a v1 query it just ignores them.  Is this the proper behavior?  I don't frankly know the difference between v1 & v2 to be honest.

**Comment by serundeputy** [comment]
I think that is fine ... if there is a want/need we can port it back to v1.

There was a need to change some endpoints for some use cases and rather than break them for any sites that are currently using them we added the functionality the v2 endpoints preserving the backwards compatibility.

Glad you got it working!

**Comment by ps23Rick** [comment]
Hmm.. If that's the case, why have a v1 interface in this particular case?  Since this is completely new functionality it seems odd, with this knowledge, to have a pair of functions that basically do the same thing.  Can I just remove my "v1" version of this?  It has zero impact on the existing code.

**Comment by ps23Rick** [comment]
Actually.. I'd rename my "v2" functionality as the v1 variant here and just not have a "v2" for this new code -- perhaps at some point that could be added IF necessary.

**Comment by serundeputy** [comment]
that seems fine to me.

**Comment by serundeputy** [comment]
that seems fine to me.

**Comment by ps23Rick** [comment]
cool.. I've got it cleaned up a bit more and removed the extraneous code that isn't wanted anymore.  

One more question -- in the case where the user has a set of say 10 items to query -- and he uses the pagination and basically goes beyond the end of the list (e.g. specifies page[size]=10 and page[number]=20 -- there's no matching data in this case and currently my code returns a code 200 (OK) but no data -- zilch.

I don't have a clue as to whether that is considered OK or should an error of some sort be kicked out in such a case..?

**Comment by ps23Rick** [comment]
I created a PR for this..

**Comment by serundeputy** [comment]
>One more question -- in the case where the user has a set of say 10 items to query -- and he uses the pagination and basically goes beyond the end of the list (e.g. specifies page[size]=10 and page[number]=20 -- there's no matching data in this case and currently my code returns a code 200 (OK) but no data -- zilch.

>I don't have a clue as to whether that is considered OK or should an error of some sort be kicked out in such a case..?

yeah; i don't know; i think a message about you've exceeded the number of pages possible for this dataset, but not those words better ones that i can't think of right now.

But, we can do that in a follow up issue/feature.

**Comment by serundeputy** [comment]
I just took a cursory glance at the PR/code it looks really good! I can't wait to try it out ... hopefully tonight I can pull it down and kick the tires.

thanks @ps23Rick !

**Comment by ps23Rick** [comment]
Thanks.. I'll admit that I'm trying to use the output of this (in my example returning 4 items from the database for my specific type) -- I'm using this in Svelte/Sapper (which is more or less JS). Unfortunately while this same code worked fine with parsing Drupal JSON data, I can't figure out why I can't dig through the data in a predictable manner.  I've checked that the data is valid JSON but what I appear to get back in an array (in my eyes) of 4 items all bundled into a JS object.  There are no names to speak of and the array is oddly setup -- my items directly from the database are NIDs 3,4,5,6.  They are accessible as data[3], data[4], data[5], and data[6].  I'm stumped as to why not data[0..3].. Anyway, I doubt this is an issue on this side of the fence but perhaps something I'm missing with JS and how it handles this sort of stuff. 

I guess the reason I ask this is that IF you see anything wonky in the output, please let me know.

I'm almost tempted to load up an instance of Drupal to compare its JSON output and see what I'm missing

**Comment by ps23Rick** [comment]
Hmm.. Ok. I was able to bring up my old Drupal install again and ultimately have it generate similar output as to what this updated code does.  There are a few things that stand out to me -- I'm not enough of a JSON junkie to know if this is bad or ok.

In the screenshot of the two -- Drupal on the left, and backdrop on the right.. Notice that "data" as Drupal calls it is called "3" on Backdrop.  That's kinda odd -- it's basically the node ID that got returned.  I *think* this is why I'm having problems in my Svelte (js) code digging through the data because it's called "3" instead of data and I can't dig through a variable specifying something like "data.3" as that's a syntax error from a JS perspective I'm fairly certain.  The other thing to note is that my query (backdrop) is returning an array and from what I've gleaned the JSON arrays ought to be bracketed as [ item1 item2 item3 ... ] but they don't get that which may be another reason JS is unhappy -- those ARE present on the Drupal dataset.  Am I missing something perhaps?  Maybe this just hasn't been noticed before? 

IF there's an issue here and it's not too in-depth, I could try to look into it..

<img width="1200" alt="Screen Shot 2020-05-21 at 4 19 42 PM" src="https://user-images.githubusercontent.com/38017680/82615512-7ab42680-9b7f-11ea-9ab0-360a59653b7a.png">

**Comment by serundeputy** [comment]
you are correct in js you can not reference `data.3` you'd need to use `data[3]`, but yeah in general we could probably get a better json data from the backdrop side.

let me pull down your PR and give a proper look. :eyes:

**Comment by ps23Rick** [comment]
Thx.. the thing that really tripped me up was that js would barf if I told it to look at data[2] which of course wasn’t present.  Hmm.. I might do some poking around in the code that generates the Json. 



Sent from myMail for iOS


Thursday, May 21, 2020, 4:48 PM -0700 from Geoff St. Pierre  <notifications@github.com>:

**Comment by ps23Rick** [comment]
Hmm.. This actually doesn't seem too bad.. It just looks like the nodes returned from backdrop just need a bit of cleanup.  What I mean by that is to re-order them inside the data structure a bit so that when backdrop_json_encode() is called that the data just comes out in a fashion that will not cause grief for those of us using JS to chew on the data.  Of course, the proofs in the puddin' so to speak.  I'll keep you posted but it appears to be still within my new code where this massaging would be tackled.  I'll go poking around in the debugger and see what I can make out.

**Comment by serundeputy** [comment]
I was just playing with your code, and doing something like this:

```php
  // Get the items to be returned in this JSON query..
  $nodes_to_return = _nodes_get_node_result($type, $opt_page_size, $opt_page_number);
  $data = [];
  foreach ($nodes_to_return as $node) {
    $data[] = $node;
  } 
  //backdrop_json_output($nodes_to_return);
  backdrop_json_output($data);
```

returns the json as an array `[]` of objects instead of an object keyed by integers, then in js you can `forEach()` over it.

but yes; you can massage it in your new/existing code

**Comment by ps23Rick** [comment]
Yes.. I tried your suggested mods and they do work, but apparently they don't go far enough.. There are similar issues I believe in deeper parts of the respective data objects.  For instance, using the above code and running it against an array of 4 sparse array items, if I pick out $data[0] and get a single instance of one of the resulting nodes, AND if I try to inspect the embedded arrays (e.g. body is one in my case), the debugger actually barfs.  In this case looking at the data in the debugger shows these occurrences of "und" (which I'd guess means undefined).. Unfortunately I'm not sure if these arrays are showing the same sort of sparseness issues or something else as my entire debug session craps out when I try to fiddle with them in PHP.

Let me know if you have any thoughts on this.. I'm thinking these occurrences of "und" are trouble.. 

<img width="476" alt="Screen Shot 2020-05-21 at 5 41 03 PM" src="https://user-images.githubusercontent.com/38017680/82619278-5f9ae400-9b8a-11ea-8a91-cb1634df722e.png">

**Comment by ps23Rick** [comment]
ok.. I think I got things ok now .. at least on the PHP side.  I was getting messed up by forgetting PHP syntax is different for array & object handling vs other languages.  I was using "." instead of "->".. That was what was crapping out the debugger.. 

Getting the field-slug on my own content type is a bit unwieldy as follows but does work in PHP:
$data[0]->field_slug['und'][0]['value'];

We'll see if this translates over to JS or not... 

I just wish I knew why the "und" is showing up in the content types.. odd..

**Comment by ps23Rick** [comment]
ok.. the js side is working just fine after I weeded through the output to get things correct. I'm happy now with the adjustments that you came up with.. They seem to do the trick quite well, but I'll give you some time to fiddle and see if there's anything I might have missed.  Thanks so much!!

**Comment by serundeputy** [comment]
sounds good.

`und` is the language index for translated sites (see `t()`).  You can get the data you need w/ `$data[0]->field_slug['und'][0]['value'];`, but what is recommended in Backdrop/Drupal is `$data[0]->field_slug[LANGUAGE_NONE][0]['value'];` that if a site has English and German it will get the correct value based on the user context; if your site is English only then `und` will work, but `LANGUAGE_NONE` will future proof it.
